src/components/postgresConnect.ts

```diff
@@ -659,6 +659,116 @@ class PostgresDB {
+    WHERE template_id = $1 
+    AND stage_number = $2
+    `
+    try {
+      const result = await this.query({ text, values: [templateId, stageNumber] })
+      return result.rows[0].max
+    } catch (err) {
+      console.log(err.message)
+      throw err
+    }
+  }
+
+  public getCurrentReviewLevel = async (reviewId: number) => {
+    const text = `SELECT level FROM review WHERE id = $1`
+    try {
+      const result = await this.query({ text, values: [reviewId] })
+      return result.rows[0].level
+    } catch (err) {
+      console.log(err.message)
+      throw err
+    }
+  }
+
+  public getReviewersForApplicationStageLevel = async (
+    templateId: number,
+    stageNumber: number,
+    reviewLevel: number
+  ) => {
+    const text = `
+    SELECT user_id, organisation_id, restrictions FROM 
+    permission_join pj JOIN template_permission tp
+    ON pj.permission_name_id = tp.permission_name_id
+    WHERE template_id = $1
+    AND stage_number = $2
+    AND level = $3
+    `
+    try {
+      const result = await this.query({
+        text,
+        values: [templateId, stageNumber, reviewLevel],
+      })
+      return result.rows
+    } catch (err) {
+      console.log(err.message)
+      throw err
+    }
+  }
+
+  public addReviewAssignments = async (reviewAssignments: any) => {
+    const reviewAssignmentIds = []
+    for (const reviewAssignment of reviewAssignments) {
+      const {
+        reviewerId,
+        orgId,
+        stageId,
+        stageNumber,
+        status,
+        applicationId,
+        templateSectionRestrictions,
+        level,
+        isLastLevel,
+      } = reviewAssignment
+      // Needs a slightly different query with different CONFLICT restrictions
+      // depending on whether orgId exists or not.
+      const text = `
+        INSERT INTO review_assignment (
+          reviewer_id,${orgId ? ' organisation_id,' : ''}
+          stage_id, stage_number, status, application_id,
+          template_section_restrictions, level, is_last_level
+          )
+        VALUES ($1, $2, $3, $4, $5, $6, $7, $8${orgId ? ', $9' : ''})
+        ON CONFLICT (reviewer_id,${
+          orgId ? ' organisation_id,' : ''
+        } stage_number, application_id, level)
+          WHERE organisation_id IS ${orgId ? 'NOT ' : ''}NULL
+        DO
+          UPDATE SET template_section_restrictions = ${orgId ? '$7' : '$6'}
+        RETURNING id`
+
+      try {
+        const result = await this.query({
+          text,
+          values: [
+            reviewerId,
+            ...(orgId ? orgId : []),
+            stageId,
+            stageNumber,
+            status,
+            applicationId,
+            templateSectionRestrictions,
+            level,
+            isLastLevel,
+          ],
+        })
+        reviewAssignmentIds.push(result.rows[0].id)
+
+        // TO-DO: What to do with existing records that don't match the
+        // generated ones? Delete them? Set their status = "Not Available"?
```

Comment:
I thought we'd just write an additional Action that regenerates the review_assignment records for a user whenever something changes for them (permissions, add/delete)? It'd be similar to this one but it would find all applications that match the user, as opposed to finding all users that match the application.

Reply:
Sure, although I was trying to focus on the comment you've made, what if permission was revoked etc.. (do we delete it, but what if it was assigned and review already in progress etc..). Seperate action would be good, but still some restrictions to what will be applied on existing applications may help tame complexity in this one.

Reply:
Okay, right. I just left this to-do in as this is where any additional checks would need to go. We'd run another query: SELECT records that match stage/application/level but with reviewers not in the list that are currently in memory. (I.e. the reviewer has probably recently been removed or permissions revoked).  And then what? Delete them? I think that would be the right thing, but probably not needed for now, right? (Hence "To-do").